crates/transpiler/src/passes/basis_translator/mod.rs

```diff
@@ -437,176 +377,185 @@ fn extract_basis_target_circ(
                 .and_modify(|set| {
                     set.insert((node_obj.op.name().to_string(), node_obj.op.num_qubits()));
                 })
-                .or_insert(IndexSet::from_iter([(
+                .or_insert(AhashIndexSet::from_iter([(
                     node_obj.op.name().to_string(),
                     node_obj.op.num_qubits(),
                 )]));
         } else {
             source_basis.insert((node_obj.op.name().to_string(), node_obj.op.num_qubits()));
         }
         if node_obj.op.control_flow() {
-            let OperationRef::Instruction(op) = node_obj.op.view() else {
-                unreachable!("Control flow op is not a control flow op. But control_flow is `true`")
-            };
-            let bound_inst = op.instruction.bind(py);
-            let blocks = bound_inst.getattr("blocks")?.try_iter()?;
-            for block in blocks {
+            for block in node_obj.op.blocks() {
                 extract_basis_target_circ(
-                    &block?,
+                    &block,
                     source_basis,
                     qargs_local_source_basis,
                     min_qubits,
                     qargs_with_non_global_operation,
-                )?;
+                );
             }
         }
     }
-    Ok(())
 }
 
 fn apply_translation(
-    py: Python,
     dag: &DAGCircuit,
-    target_basis: &IndexSet<String, ahash::RandomState>,
+    target_basis: &AhashIndexSet<&str>,
     instr_map: &InstMap,
     extra_inst_map: &ExtraInstructionMap,
     min_qubits: usize,
-    qargs_with_non_global_operation: &IndexMap<
-        Qargs,
-        IndexSet<String, ahash::RandomState>,
-        ahash::RandomState,
-    >,
-) -> PyResult<(DAGCircuit, bool)> {
+    qargs_with_non_global_operation: &AhashIndexMap<Qargs, AhashIndexSet<&str>>,
+) -> Result<(DAGCircuit, bool), BasisTranslatorError> {
     let mut is_updated = false;
-    let out_dag = dag.copy_empty_like(VarsMode::Alike)?;
+    let out_dag = dag
+        .copy_empty_like(VarsMode::Alike)
+        .map_err(|err| BasisTranslatorError::ApplyTranslationCircuitError(err.to_string()))?;
```

Comment:
What was the conclusion on calling `err.to_string()` now? Should we just map this to a new error without calling `to_string`? E.g.
```suggestion
        .map_err(|_| BasisTranslatorError::TranslationFailed)?;
```

Reply:
I just made generic `Circuit`, and `DAGCircuit` errors that may be replaced in the future when we implement rust native errors for the rest of the API. Once that's done we can change the errors to use `#[from] CircuitError` when that is done. I'll add `TODO` notes.

Reply:
Added in baabe1c1fc8d03bfbf2be07418f4799d44bb800e